File: typescript/src/schema/insights.ts
import { z } from "zod";
import { InsightQuerySchema } from "./query";

export const InsightSchema = z.object({
	id: z.number(),
	name: z.string().nullish(),
	description: z.string().nullish(),
	filters: z.record(z.any()),
	query: z.record(z.any()).nullish(),
	result: z.any().optional(),
	created_at: z.string(),
	updated_at: z.string(),
	created_by: z
		.object({
			id: z.number(),
			uuid: z.string().uuid(),
			distinct_id: z.string(),
			first_name: z.string(),
			email: z.string(),
		})
		.optional()
		.nullable(),
	saved: z.boolean(),
	favorited: z.boolean().nullish(),
	deleted: z.boolean(),
	dashboard: z.number().nullish(),
	layouts: z.record(z.any()).nullish(),
	color: z.string().nullish(),
	last_refresh: z.string().nullish(),
	refreshing: z.boolean().nullish(),
	tags: z.array(z.string()).nullish(),
});

export const CreateInsightInputSchema = z.object({
	name: z.string(),
	query: InsightQuerySchema,
	description: z.string().optional(),
	saved: z.boolean().default(true),
	favorited: z.boolean().default(false),
	tags: z.array(z.string()).optional(),
});

export const UpdateInsightInputSchema = z.object({
	name: z.string().optional(),
	description: z.string().optional(),
	filters: z.record(z.any()).optional(),
	query: z.record(z.any()).optional(),
	saved: z.boolean().optional(),
	favorited: z.boolean().optional(),
	dashboard: z.number().optional(),
	tags: z.array(z.string()).optional(),
});

export const ListInsightsSchema = z.object({
	limit: z.number().optional(),
	offset: z.number().optional(),
	saved: z.boolean().optional(),
	favorited: z.boolean().optional(),
	search: z.string().optional(),
});

export type PostHogInsight = z.infer<typeof InsightSchema>;
export type CreateInsightInput = z.infer<typeof CreateInsightInputSchema>;
export type UpdateInsightInput = z.infer<typeof UpdateInsightInputSchema>;
export type ListInsightsData = z.infer<typeof ListInsightsSchema>;

export const SQLInsightResponseSchema = z.array(
	z.object({
		type: z.string(),
		data: z.record(z.any()),
	}),
);

export type SQLInsightResponse = z.infer<typeof SQLInsightResponseSchema>;

File: python/schema/insights.py
from typing import Any
from uuid import UUID

from pydantic import BaseModel, RootModel

from schema.query import InsightQuery


class User(BaseModel):
    id: int
    uuid: UUID
    distinct_id: str
    first_name: str
    email: str


class Insight(BaseModel):
    id: int
    name: str
    description: str | None = None
    filters: dict[str, Any]
    query: dict[str, Any] | None = None
    result: Any | None = None
    created_at: str
    updated_at: str
    created_by: User | None = None
    saved: bool
    favorited: bool | None = None
    deleted: bool
    dashboard: int | None = None
    layouts: dict[str, Any] | None = None
    color: str | None = None
    last_refresh: str | None = None
    refreshing: bool | None = None
    tags: list[str] | None = None


class CreateInsightInput(BaseModel):
    name: str
    query: InsightQuery
    description: str | None = None
    saved: bool = True
    favorited: bool = False
    tags: list[str] | None = None


class UpdateInsightInput(BaseModel):
    name: str | None = None
    description: str | None = None
    filters: dict[str, Any] | None = None
    query: InsightQuery | None = None
    saved: bool | None = None
    favorited: bool | None = None
    dashboard: int | None = None
    tags: list[str] | None = None


class ListInsights(BaseModel):
    limit: int | None = None
    offset: int | None = None
    saved: bool | None = None
    favorited: bool | None = None
    search: str | None = None


class SQLInsightResponseItem(BaseModel):
    type: str
    data: dict[str, Any]


class SQLInsightResponse(RootModel[list[SQLInsightResponseItem]]):
    root: list[SQLInsightResponseItem]
